File: src/2025-02-23-deepclaude-try/java/exercises/practice/circular-buffer/src/main/java/CircularBuffer.java
class BufferIOException extends RuntimeException {
    BufferIOException(String message) {
        super(message);
    }
}

class CircularBuffer<T> {

    private final T[] buffer;
    private int readIndex = 0;
    private int writeIndex = 0;
    private int count = 0;
    
    @SuppressWarnings("unchecked")
    CircularBuffer(final int size) {
        buffer = (T[]) new Object[size];
    }

    T read() throws BufferIOException {
        if (count == 0) {
            throw new BufferIOException("Cannot read from empty buffer");
        }
        T value = buffer[readIndex];
        readIndex = (readIndex + 1) % buffer.length;
        count--;
        return value;
    }

    void write(T data) throws BufferIOException {
        if (count == buffer.length) {
            throw new BufferIOException("Cannot write to full buffer");
        }
        buffer[writeIndex] = data;
        writeIndex = (writeIndex + 1) % buffer.length;
        count++;
    }

    void overwrite(T data) {
        if (count == buffer.length) {
            readIndex = (readIndex + 1) % buffer.length;
            count--;
        }
        buffer[writeIndex] = data;
        writeIndex = (writeIndex + 1) % buffer.length;
        count++;
    }

    void clear() {
        readIndex = 0;
        writeIndex = 0;
        count = 0;
    }

}

File: src/2025-02-23-deepclaude-try/cpp/exercises/practice/circular-buffer/circular_buffer.cpp
#include "circular_buffer.h"

namespace circular_buffer {

circular_buffer::circular_buffer(size_t capacity) 
    : capacity_(capacity), data_(capacity) {}

int circular_buffer::read() {
    if (is_empty()) {
        throw std::runtime_error("Buffer is empty");
    }
    int value = data_[start_];
    start_ = (start_ + 1) % capacity_;
    --size_;
    return value;
}

void circular_buffer::write(int value) {
    if (is_full()) {
        throw std::runtime_error("Buffer is full");
    }
    data_[end_] = value;
    end_ = (end_ + 1) % capacity_;
    ++size_;
}

void circular_buffer::overwrite(int value) {
    if (is_full()) {
        start_ = (start_ + 1) % capacity_;
        --size_;
    }
    data_[end_] = value;
    end_ = (end_ + 1) % capacity_;
    ++size_;
}

bool circular_buffer::is_empty() const {
    return size_ == 0;
}

bool circular_buffer::is_full() const {
    return size_ == capacity_;
}

}  // namespace circular_buffer
